pkg/controller/registry/resolver/sat/dict.go

```diff
@@ -0,0 +1,268 @@
+}
+
+// compileDict returns a new dict with its state initialized based on
+// the provided slice of Installables. This includes construction of
+// the translation tables between Installables/Constraints and the
+// inputs to the underlying solver.
+func compileDict(installables []Installable) *dict {
+	d := dict{
+		installables: make(map[z.Lit]Installable, len(installables)),
+		lits:         make(map[Identifier]z.Lit, len(installables)),
+		constraints:  make(map[z.Lit]AppliedConstraint),
+		c:            logic.NewC(),
+	}
+
+	var buf []z.Lit
+	var x cstate
+	weights := make([]z.Lit, 0, len(installables))
+
+	// First pass to assign lits from the circuit:
+	for _, installable := range installables {
+		im := d.c.Lit()
+		d.installables[im] = installable
+		d.lits[installable.Identifier()] = im
+		weights = append(weights, im)
+	}
+
+	// Then build the constraints:
+	for _, installable := range installables {
+		for _, constraint := range installable.Constraints() {
+			x.Reset()
+			constraint.apply(&x, installable.Identifier())
+
+			clauses := make([]z.Lit, 0, x.weight)
+			for w := 0; w < x.weight; w++ {
+				wm := d.c.Lit()
+				weights = append(weights, wm)
+				clauses = append(clauses, d.c.Implies(d.LitOf(installable.Identifier()), wm))
+			}
+
+			if !x.Empty() {
+				buf = buf[:0]
+				for _, p := range x.pos {
+					buf = append(buf, d.LitOf(p))
+				}
+				for _, n := range x.neg {
+					buf = append(buf, d.LitOf(n).Not())
+				}
+				clauses = append(clauses, d.c.Ors(buf...))
+			}
+			m := d.c.Ands(clauses...)
+
+			d.constraints[m] = AppliedConstraint{
+				Installable: installable,
+				Constraint:  constraint,
+			}
+		}
+	}
+
+	d.cards = d.c.CardSort(weights)
+
+	return &d
+}
+
+// Solve searches for a solution to the problem encoded in a dict that
+// minimizes the total weight of all selected Installables.
+func (d *dict) Solve(ctx context.Context, g *gini.Gini) int {
+	d.c.ToCnf(g)
+
+	var result int
+	linearSearch(0, d.cards.N(), func(w int) bool {
+		if w >= 0 {
+			// Omitting the cardinality constraint on the
+			// last iteration prevents the cardinality
+			// constraint literal from showing up in the
+			// set of failed assumptions should there be
+			// no solution.
+			g.Assume(d.cards.Leq(w))
+		}
+		for m := range d.constraints {
+			g.Assume(m)
+		}
+		result = waitForSolution(ctx, g.GoSolve())
+		return result == satisfiable
+	})
+
+	return result
+}
+
+func waitForSolution(ctx context.Context, gs inter.Solve) int {
+	t := time.NewTicker(50 * time.Millisecond)
+	defer t.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return gs.Stop()
+		case <-t.C:
+			if result, ok := gs.Test(); ok {
+				return result
+			}
+		}
+	}
+}
+
+func linearSearch(min, max int, f func(int) bool) {
+	for x := min; x <= max; x++ {
+		if f(x) {
+			return
+		}
+	}
+	f(-1) // omit cardinality constraint
+}
+
+func binarySearch(min, max int, f func(int) bool) {
+	for {
+		x := (min + max) / 2
```

Comment:
could add a possible check for [overflow](https://stackoverflow.com/questions/6735259/calculating-mid-in-binary-search)?

Reply:
Sure, although if the input size is anywhere near to overflowing min+max, we'll have other problems. Good catch.